libs/sdk-core/src/greenlight/node_api.rs

```diff
@@ -704,6 +715,61 @@ impl Greenlight {
         res
     }
 
+    async fn with_connection_fallback<T, M, F>(
+        &self,
+        main: M,
+        fallback: impl FnOnce() -> F,
+    ) -> Result<T, tonic::Status>
+    where
+        M: Future<Output = Result<T, tonic::Status>>,
+        F: Future<Output = Result<T, tonic::Status>>,
+        T: std::fmt::Debug,
+    {
+        let res = main.await;
+        let status = match res {
+            Ok(t) => return Ok(t),
+            Err(s) => s,
+        };
+        info!("AAAAAAAAAAAAAAA got status");
+
+        let source = match status.source() {
+            Some(source) => source,
+            None => return Err(status),
+        };
+        info!("AAAAAAAAAAAAAAA got source");
+
+        let error: &tonic::transport::Error = match source.downcast_ref() {
+            Some(error) => error,
+            None => return Err(status),
+        };
+        info!("AAAAAAAAAAAAAAA got error");
+
+        if error.to_string() != "transport error" {
+            return Err(status);
+        }
+        info!("AAAAAAAAAAAAAAA is transport error");
+
+        let source = match error.source() {
+            Some(source) => source,
+            None => return Err(status),
+        };
+        info!("AAAAAAAAAAAAAAA got source");
+        info!("AAAAAAAAAAAAAAA {}", source.to_string());
+        if !source.to_string().contains("keep-alive timed out") {
+            return Err(status);
+        }
+
+        info!("AAAAAAAAAAAAAAA is connection error");
+        if self.network_change_detector.send(()).is_err() {
+            warn!("greenlight network change detector died.");
+        }
+
+        let res = fallback().await;
```

Comment:
Do we need to wait at all to allow time for the GRPC clients to reconnect?

Reply:
No it looks like the connection is immediately reusable.

You can test this with the sdk-cli by:
- connecting the sdk
- changing your laptop to another network
- calling send_payment or lnurl_pay with or without --use_trampoline

It will hang for a long time (until this PR https://github.com/Blockstream/greenlight/pull/548 is included) and then continue with the fallback successfully.